File: sdks/nodejs/examples/caching.ts
/**
 * Caching Example - TollMeshCache Node.js SDK
 */

import { Client, ClientConfig } from '../dist/client';

async function main() {
  const config: ClientConfig = {
    host: 'localhost',
    port: 8080,
  };

  const client = new Client(config);

  console.log('='.repeat(60));
  console.log('TollMeshCache - Caching Example');
  console.log('='.repeat(60));

  try {
    // Example 1: Simple cache operations
    console.log('\n1. Simple Cache Operations');
    console.log('-'.repeat(60));

    const namespace = 'users';
    const key = 'user-123';
    const value = JSON.stringify({
      id: '123',
      name: 'Alice',
      email: 'alice@example.com',
    });

    console.log(`Setting ${namespace}/${key}...`);
    await client.cacheSet(namespace, key, value, 3600000);
    console.log('✓ Set successful');

    console.log(`Getting ${namespace}/${key}...`);
    const { value: cachedValue, exists } = await client.cacheGet(namespace, key);
    if (exists) {
      console.log(`✓ Found: ${cachedValue}`);
    } else {
      console.log('✗ Not found');
    }

    // Example 2: Cache-aside pattern
    console.log('\n2. Cache-Aside Pattern (Lazy Loading)');
    console.log('-'.repeat(60));

    async function getUser(userId: string) {
      const ns = 'user_profiles';
      console.log(`Checking cache for ${userId}...`);
      const { value, exists } = await client.cacheGet(ns, userId);

      if (exists) {
        console.log('✓ Cache hit!');
        return JSON.parse(value!);
      }

      console.log('✗ Cache miss - fetching from source...');
      const userData = {
        id: userId,
        name: `User ${userId}`,
        email: `user${userId}@example.com`,
      };

      console.log('  Caching result (1 hour TTL)...');
      await client.cacheSet(ns, userId, JSON.stringify(userData), 3600000);

      return userData;
    }

    console.log('\nFirst call to getUser("user-456"):');
    const user1 = await getUser('user-456');
    console.log(`Result: ${user1.name}`);

    console.log('\nSecond call to getUser("user-456"):');
    const user2 = await getUser('user-456');
    console.log(`Result: ${user2.name}`);

    // Example 3: Multiple namespaces
    console.log('\n3. Multiple Cache Namespaces');
    console.log('-'.repeat(60));

    const namespaces = {
      users: { ttl: 3600000, key: 'user-789' },
      posts: { ttl: 86400000, key: 'post-101' },
      comments: { ttl: 21600000, key: 'comment-202' },
      sessions: { ttl: 3600000, key: 'session-abc123' },
    };

    for (const [ns, config] of Object.entries(namespaces)) {
      const data = { namespace: ns, timestamp: new Date().toISOString() };
      await client.cacheSet(ns, config.key, JSON.stringify(data), config.ttl);
      console.log(`✓ Cached ${ns}/${config.key} (TTL: ${config.ttl}ms)`);
    }

    // Example 4: Different data types
    console.log('\n4. Caching Different Data Types');
    console.log('-'.repeat(60));

    const testCases = [
      { ns: 'strings', key: 'greeting', value: 'Hello, World!' },
      { ns: 'numbers', key: 'pi', value: JSON.stringify(3.14159) },
      { ns: 'objects', key: 'person', value: JSON.stringify({ name: 'Bob', age: 30 }) },
      { ns: 'arrays', key: 'scores', value: JSON.stringify([95, 87, 92, 88, 91]) },
    ];

    for (const { ns, key: k, value: v } of testCases) {
      await client.cacheSet(ns, k, v, 3600000);
      const { value: cachedVal } = await client.cacheGet(ns, k);
      console.log(`✓ ${ns.padEnd(10)}/${k.padEnd(10)}: ${cachedVal?.substring(0, 50)}`);
    }

    // Example 5: Cache statistics
    console.log('\n5. Server Statistics');
    console.log('-'.repeat(60));

    const health = await client.health();
    console.log(`Cache Status: ${health.status}`);
    if (health.stats) {
      console.log(`Cache Hit Rate: ${(health.stats.cache_hit_rate * 100).toFixed(1)}%`);
      console.log(`Total Requests: ${health.stats.requests_total}`);
    }

  } catch (error) {
    console.error('ERROR:', error);
  } finally {
    client.close();
    console.log('\n' + '='.repeat(60));
    console.log('Example complete!');
    console.log('='.repeat(60));
  }
}

main();

File: sdks/python/examples/caching.py
#!/usr/bin/env python3
"""
Caching Example - TollMeshCache Python SDK

Demonstrates how to use TollMeshCache for distributed caching.
"""

import json
from datetime import timedelta
import sys
sys.path.insert(0, '../')

from tollmeshcache import Client, ClientConfig, TollMeshError


def simulate_database_query(user_id: str) -> dict:
    """Simulate an expensive database query"""
    print(f"    [DB] Fetching user {user_id} from database...")
    return {
        "id": user_id,
        "name": f"User {user_id}",
        "email": f"user{user_id}@example.com",
        "created_at": "2024-01-01T00:00:00Z"
    }


def simulate_api_call(endpoint: str) -> dict:
    """Simulate an expensive API call"""
    print(f"    [API] Calling {endpoint}...")
    return {
        "status": "success",
        "data": [1, 2, 3, 4, 5],
        "timestamp": "2024-01-01T00:00:00Z"
    }


def main():
    config = ClientConfig(
        host="localhost",
        port=8080,
        timeout=5.0,
    )

    client = Client(config)

    print("=" * 60)
    print("TollMeshCache - Distributed Caching Example")
    print("=" * 60)

    try:
        # Example 1: Simple cache get/set
        print("\n1. Simple Cache Operations")
        print("-" * 60)

        namespace = "users"
        key = "user-123"
        value = json.dumps({
            "id": "123",
            "name": "Alice",
            "email": "alice@example.com"
        })

        print(f"Setting {namespace}/{key}...")
        client.cache_set(namespace, key, value, ttl=timedelta(hours=1))
        print("✓ Set successful")

        print(f"Getting {namespace}/{key}...")
        cached_value, exists = client.cache_get(namespace, key)
        if exists:
            print(f"✓ Found: {cached_value}")
        else:
            print("✗ Not found")

        # Example 2: Cache-aside pattern (lazy loading)
        print("\n2. Cache-Aside Pattern (Lazy Loading)")
        print("-" * 60)

        def get_user(user_id: str) -> dict:
            """Get user with caching"""
            namespace = "user_profiles"
            key = user_id

            # Try cache first
            print(f"Checking cache for {user_id}...")
            cached_value, exists = client.cache_get(namespace, key)

            if exists:
                print("✓ Cache hit!")
                return json.loads(cached_value)

            # Cache miss - fetch from database
            print("✗ Cache miss - fetching from source...")
            user_data = simulate_database_query(user_id)

            # Store in cache
            print(f"  Caching result (1 hour TTL)...")
            client.cache_set(
                namespace,
                key,
                json.dumps(user_data),
                ttl=timedelta(hours=1)
            )

            return user_data

        # First call - cache miss
        print("\nFirst call to get_user('user-456'):")
        user1 = get_user("user-456")
        print(f"Result: {user1['name']}")

        # Second call - cache hit
        print("\nSecond call to get_user('user-456'):")
        user2 = get_user("user-456")
        print(f"Result: {user2['name']}")

        # Example 3: Multiple cache namespaces
        print("\n3. Multiple Cache Namespaces")
        print("-" * 60)

        namespaces = {
            "users": {"ttl": timedelta(hours=1), "sample_key": "user-789"},
            "posts": {"ttl": timedelta(days=1), "sample_key": "post-101"},
            "comments": {"ttl": timedelta(hours=6), "sample_key": "comment-202"},
            "sessions": {"ttl": timedelta(hours=1), "sample_key": "session-abc123"},
        }

        for ns, config in namespaces.items():
            data = {"namespace": ns, "timestamp": "2024-01-01"}
            client.cache_set(
                ns,
                config["sample_key"],
                json.dumps(data),
                ttl=config["ttl"]
            )
            print(f"✓ Cached {ns}/{config['sample_key']} (TTL: {config['ttl']})")

        # Example 4: Cache invalidation
        print("\n4. Cache Invalidation Patterns")
        print("-" * 60)

        # Store some data
        namespace = "config"
        key = "app_settings"
        config_data = {"theme": "dark", "language": "en"}

        client.cache_set(namespace, key, json.dumps(config_data), ttl=timedelta(hours=1))
        print(f"✓ Cached {namespace}/{key}")

        # Verify it's there
        value, exists = client.cache_get(namespace, key)
        print(f"✓ Retrieved: {value}")

        # Invalidate by setting with 0 TTL (or deleting)
        print("Invalidating cache (not yet supported - manual cleanup on update)...")

        # Update and recache
        config_data["theme"] = "light"
        client.cache_set(namespace, key, json.dumps(config_data), ttl=timedelta(hours=1))
        print(f"✓ Updated and recached {namespace}/{key}")

        # Example 5: Cache different data types
        print("\n5. Caching Different Data Types")
        print("-" * 60)

        test_cases = [
            ("strings", "greeting", "Hello, World!"),
            ("numbers", "pi", json.dumps(3.14159)),
            ("objects", "person", json.dumps({"name": "Bob", "age": 30})),
            ("arrays", "scores", json.dumps([95, 87, 92, 88, 91])),
        ]

        for namespace, key, value in test_cases:
            client.cache_set(namespace, key, value, ttl=timedelta(hours=1))
            cached_val, exists = client.cache_get(namespace, key)
            print(f"✓ {namespace:10s}/{key:10s}: {cached_val[:50]}")

        # Example 6: Cache statistics
        print("\n6. Server Statistics")
        print("-" * 60)

        try:
            health = client.health()
            print(f"Cache Status: {health['status']}")
            if 'stats' in health:
                stats = health['stats']
                print(f"Cache Hit Rate: {stats.get('cache_hit_rate', 0):.1%}")
                print(f"Total Requests: {stats.get('requests_total', 0)}")
        except TollMeshError as e:
            print(f"Health check error: {e.message}")

    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
    except Exception as e:
        print(f"\nFATAL ERROR: {e}")
        import traceback
        traceback.print_exc()
    finally:
        client.close()
        print("\n" + "=" * 60)
        print("Example complete!")
        print("=" * 60)


if __name__ == "__main__":
    main()
